solarwinds_apm/oboe/sampler.py

```diff
@@ -0,0 +1,309 @@
+            self._tracing_mode = None
+        self._trigger_mode = config.trigger_trace_enabled
+        self._transaction_settings = config.transaction_settings
+        self._ready = threading.Event()
+        if initial:
+            self.update_settings(initial)
+
+    def __str__(self) -> str:
+        return f"Sampler{self._tracing_mode}({self._trigger_mode}) {super().__str__(self)}"
+
+    @property
+    def tracing_mode(self):
+        return self._tracing_mode
+
+    @property
+    def trigger_mode(self):
+        return self._trigger_mode
+
+    @property
+    def transaction_settings(self):
+        return self._transaction_settings
+
+    def wait_until_ready(self, timeout: int) -> bool:
+        """
+        Waits until the sampler is ready.
+        """
+        return self._ready.wait(timeout)
+
+    @override
+    def local_settings(
+        self,
+        parent_context: "Context" | None,
+        trace_id: int,
+        name: str,
+        kind: SpanKind | None = None,
+        attributes: Attributes = None,
+        links: Sequence["Link"] | None = None,
+        trace_state: "TraceState" | None = None,
+    ) -> LocalSettings:
+        """
+        Returns local settings.
+        """
+        settings = LocalSettings(
+            tracing_mode=self.tracing_mode, trigger_mode=self.trigger_mode
+        )
+        if (
+            self.transaction_settings is None
+            or len(self.transaction_settings) == 0
+        ):
+            return settings
+        meta = http_span_metadata(kind, attributes)
+        identifier = (
+            meta["url"] if meta["http"] else f"{SpanKind(kind).name}:{name}"
+        )
+        for transaction_setting in self.transaction_settings:
+            if transaction_setting.matcher and transaction_setting.matcher(
+                identifier
+            ):
+                settings.tracing_mode = (
+                    TracingMode.ALWAYS
+                    if transaction_setting.tracing
+                    else TracingMode.NEVER
+                )
+                break
+        return settings
+
+    @override
+    def request_headers(
+        self,
+        parent_context: "Context" | None,
+        trace_id: int,
+        name: str,
+        kind: SpanKind | None = None,
+        attributes: Attributes = None,
+        links: Sequence["Link"] | None = None,
+        trace_state: "TraceState" | None = None,
+    ) -> RequestHeaders:
+        """
+        Returns request headers.
+        """
+        if parent_context:
+            options = parent_context.get(INTL_SWO_X_OPTIONS_KEY)
+            if options and isinstance(options, XTraceOptions):
+                RequestHeaders(
+                    x_trace_options=options.options_header,
+                    x_trace_options_signature=options.signature,
+                )
+        return RequestHeaders(
+            x_trace_options=None, x_trace_options_signature=None
+        )
+
+    @override
+    def set_response_headers(
```

Comment:
Is this a replacement for the [custom ResponsePropagator's inject](https://github.com/solarwinds/apm-python/blob/4caf5852935f3b32bdccf0bf1206f5132bd16291/solarwinds_apm/response_propagator.py#L39-L45) function? Which trigger trace scenarios were tested wtih HttpSampler?

Reply:
After learning the Propagator API, the `set_response_headers` is not a replacement for the propagator inject function. 
This function is trying to set the x-trace-options-response back to the parent context so that the propagator inject function can retrieve the response from the parent context to inject the carrier.

Reply:
Awesome thank you